gapir/cc/stack.h

```diff
@@ -26,457 +26,172 @@
+        mStack(new Entry[size]),
+        mMemoryManager(memoryManager) {}
+
+  ~Stack() { delete[] mStack; }
+
   template <typename T>
-  T pop() {
-    if (!popCheck("pop")) {
-      return T();
+  T pop(typename std::enable_if<std::is_pointer<T>::value>::type* = 0) {
+    assert(mNumEntries > 0);
+
+    if (isVolatilePointerAtTop()) {
+      VolatilePointer volPtr = mStack[--mNumEntries].get<VolatilePointer>();
+      return (T)mMemoryManager->volatileToAbsolute(volPtr.getOffset());
+    } else if (typeAtTop() == typeid(ConstantPointer)) {
+      ConstantPointer constPtr = mStack[--mNumEntries].get<ConstantPointer>();
+      return (T)mMemoryManager->constantToAbsolute(constPtr.getOffset());
     }
-    GAPID_VERBOSE("-%s pop()", mStack[mTop - 1].debugInfo(mMemoryManager));
-    return PopImpl<T>::pop(this);
+
+    return (T)mStack[--mNumEntries].get<void*>();
+  }
+
+  template <typename T>
+  T pop(typename std::enable_if<!std::is_pointer<T>::value>::type* = 0) {
+    assert(mNumEntries > 0);
+    return mStack[--mNumEntries].get<T>();
   }
 
-  // Pop the static array (stored as a pointer) from the top of the stack.
   template <typename T, int N>
   core::StaticArray<T, N> pop() {
     T* ptr = pop<T*>();
-    core::StaticArray<T, N> out;
+    core::StaticArray<T, N> ret;
     for (int i = 0; i < N; i++) {
-      out[i] = ptr[i];
+      ret[i] = ptr[i];
     }
-    return out;
+    return ret;
   }
 
-  // Pop the volatile pointer from the top of the stack. If the top element
-  // is not a volatile pointer puts the stack into and invalid state. Also,
-  // puts the stack into an invalid state if called on an empty stack.
-  // Use with care, this template casts a pointer to a T* with no way of
-  // knowing if that is safe.
   template <typename T>
-  T* popVolatile() {
-    if (!popCheck("popVolatile")) {
-      return nullptr;
-    }
-    BaseType type = getTopType();
-    if (type != BaseType::VolatilePointer) {
-      GAPID_WARNING("popVolatile called for type %s", baseTypeName(type));
-      mValid = false;
-      return nullptr;
-    }
-    return pop<T*>();
+  void push(T value) {
+    assert(mNumEntries < mMaxEntries);
+    mStack[mNumEntries++].set(value);
   }
 
-  // Pop the constant pointer from the top of the stack. If the top element
-  // is not a constant pointer puts the stack into and invalid state. Also,
-  // puts the stack into an invalid state if called on an empty stack.
-  // Use with care, this template casts a pointer to a const T* with no way of
-  // knowing if that is safe.
-  template <typename T>
-  const T* popConstant() {
-    if (!popCheck("popConstant")) {
-      return nullptr;
-    }
-    BaseType type = getTopType();
-    if (type != BaseType::ConstantPointer) {
-      GAPID_WARNING("popConstant called for type %s", baseTypeName(type));
-      mValid = false;
-      return nullptr;
-    }
-    return pop<const T*>();
-  }
+  void discard(uint32_t count) { assert(mNumEntries-- > 0); }
```

Comment:
This is ignoringthe (count) parameter.

Reply:
Fixed. Good catch! I've been sick this week :(